### ISSUE TITLE
e621 plugin does not respect blacklisted tags for open all in tabs

### ISSUE THREAD

**Issue by Simon-Tesla** [Simon-Tesla/RaccoonyWebEx#122]
**Describe the bug**
e621 hides blacklisted items using CSS but doesn't remove them from the DOM so they're still picked up by the plugin.

**To Reproduce**
1. Blacklist a tag
2. Search for that tag; observe the page is blank
3. Choose "Open all in tabs"

**Expected behavior**
No tabs are opened; ideally raccoony won't show the option if there are no visible submissions on a page.

**Comment by Eupeptic1** [comment]
Can reproduce with Raccoony 1.2.5 and Firefox 78.15.0esr, 64-bit, on Debian Linux, 64-bit.

#### Analysis

Blacklisted submissions (eventually) have `blacklisted-active` as part of their class.  I think we should key off of that, BUT I think we need to somehow wait for the CSS to happen first, because that string doesn't seem to be in the class "at the beginning of the world".

(We shouldn't key off of just `blacklisted`, because that appears in the class for submissions that aren't blacklisted - I don't know why.)

#### Attempts to fix

I first tried adding various things involving `:not(.blacklisted-active)` to the query selector on [line 202](
https://github.com/Simon-Tesla/RaccoonyWebEx/blob/9ad087cf6777e21574ef9eecc5d577e5857b7539/app/plugins/e621.ts#L202) of e621.ts.  None of them worked; I initially figured i was screwing up the syntax.

I then tried adding a filter to that selector, still picking the anchor elements, and then going up one level to see if the parent's class had `blacklisted-active` in it:

        let links: HTMLAnchorElement[] = (querySelectorAll("#posts-container .post-preview a") as HTMLAnchorElement[])
            .filter((elt) => {
                // Filter out any blacklisted submissions
                logger.log("e621: element ", elt); 
                logger.log("e621: parent element ", elt.parentElement);
                logger.log("e621: parent element class list", elt.parentElement.classList);
                logger.log("e621: want to filter? ", elt.parentElement.classList.contains("blacklisted-active"));
                logger.log("e621: parent element class list value", elt.parentElement.classList.value);
                logger.log("e621: want to filter? ", elt.parentElement.classList.value.indexOf("blacklisted-active"));
                return (elt.parentElement.classList.contains("blacklisted-active"))
            });
        logger.log("e621: raw links", links);

That didn't work either; both the `classList` and `value` never had `blacklisted-active` in them, so the "want to filter?" prints always said false.

I *suspect* the e621 plugin is seeing the DOM before the CSS fools with it.  Partly that's because the "class list" and "class list value" prints in the above code only show four things (`post-preview captioned post-pos-score post-rating-explicit`), and the "class list" print makes Firefox's console say it is a `DomTokenList(4)`.  However, when I expand that `DomTokenList(4)` in the console, 1) it actually has six array elements - `blacklisted` and `blacklisted-active` are the new ones; 2) it has `length` of 6; 3) `value` includes those two new elements as well.

If the answer is to wait for the CSS, all of the above code can *probably* be replaced by a `:not()` with the correct syntax in the existing selector, like I initially tried.

**Comment by Simon-Tesla** [comment]
Given that it isn't filtered on the server side, it's almost definitely being handled by some client-side JS adding the class. Not super elegant, but a great way to move something computationally expensive to the client.

**Comment by Eupeptic1** [comment]
> Not super elegant, but a great way to move something computationally expensive to the client.

And if e621 knows anything, it's how to move something expensive to their clients!  >_>

#### Further analysis

The script at e621 that does the deed seems to be https://e621 dot net/packs/js/application-d13bfa1146bab2b6755f.js .  (The hash probably changes over time, but it's the only one named application-[hash].js.)

That script appears to add the `blacklisted` class to *all* the submissions on a search page, just to show that the blacklist process has run.  If a submission is actually blacklisted, it *also* gets the `blacklisted-active` class added.

#### Possible fix

I added a mutation selector on `body` .

I changed the current [line 202](https://github.com/Simon-Tesla/RaccoonyWebEx/blob/9ad087cf6777e21574ef9eecc5d577e5857b7539/app/plugins/e621.ts#L202) to

    let links: HTMLAnchorElement[] = querySelectorAll("#posts-container .post-preview.blacklisted:not(.blacklisted-active) a") as HTMLAnchorElement[];

and then I check the size of `links`.  If it's zero, I bail out of `getMedia()`, and hope for better results the next time the mutation selector calls us.

If none of the submissions have `.blacklisted` in the class, the blacklist script hasn't run yet, so bailing out is OK.  If all of the submissions have both `.blacklisted` AND `.blacklisted-active` in the class, then everything on the page is blacklisted, and bailing out is also OK.  If the submissions have `.blacklisted` in the class, and some of them don't have `.blacklisted-active` in the class, then that's the list we want to present for open-in-tabs.

#### Testing

It seems to work OK, in limited testing.  If there is a mix of non-blacklisted and blacklisted submissions on the page, the blacklisted submissions do not appear in the open-in-tabs list.  If all of the submissions on the page are blacklisted, the raccoon head does not appear at all.

If one clicks the "disable all" link under the "Blacklisted" heading in the left side bar, the plugin runs again, and includes the newly-visible posts in the open-in-tabs list, along with the posts that were previously visible.

If nothing on the page is blacklisted, everything on the page makes it to the open-all-in-tabs list.

It works the same when you are not logged in (tested with tags on the default blacklist) as when you are logged in (tested with tags on my personal blacklist).

The plugin seems to get called twice per search-results page, regardless if anything on the page is blacklisted or not.

#### Availability

Proposed fix is on the [e621-blacklist branch](https://github.com/Eupeptic1/RaccoonyWebEx/tree/e621-blacklist) in my repo.  Only the e621 plugin itself changed.

**Comment by Eupeptic1** [comment]
### Question:

As filed, this is about submissions that appear on the search page.

Do we *also* want this to apply to the pages for individual submissions?  In other words, if a submission is blacklisted, do we want Raccoony not to run (no raccoon head) on that page?

Point in favor: Makes Raccoony more consistent about blacklisted submissions; they are ignored on the search page *and* on the individual submission page.

Points against: e621 itself still lets you download the original image (via their "download" button), even if the submission is blacklisted, so maybe we should to.  May be confusing to users if Raccoony doesn't appear to be running on some submission pages.  (Then again, maybe they will correctly assume it's nor running because the image is blacklisted.)

**Comment by Eupeptic1** [comment]
### The proposed fix is not ready yet.

I was dogfooding and found at least one search where the proposed fix fails to find anything: https://e621 dot net/posts?tags=helvetica_(artist)

None of the posts on that page ever get the `.blacklisted` in the class, showing that the blacklist process has run, so the plugin never finds anything, and the raccoon head doesn't display.

I'm not sure why the posts don't get that added to their class.  I don't *think* it has to do with being logged in to e621 or not, because I tried it both ways, and I didn't get the raccoon head either way.

None of the posts on that search appear to be on the e621 default blacklist, or my personal blacklist, so maybe that's why... maybe the blacklist Javascript at e621 doesn't add the `.blacklisted` classes if nothing on the page needs it?  Not sure yet.

**Comment by Simon-Tesla** [comment]
I'd prefer Raccoony to always be available on an image submission so long as the image is being displayed to the user, otherwise it looks like a bug. The point of the blacklist filtering is to avoid showing people stuff that they don't want to see, not prevent them from saving it once they've bypassed the blacklist to look at it anyway.

It might make sense to generate an initial list and then check for the blacklisted class on the [element.classList property](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList) on that initial list of elements and filter them out if it's present. Doing clever things with CSS selectors isn't something I like to do in any case.

**Comment by Eupeptic1** [comment]
> I'd prefer Raccoony to always be available on an image submission so long as the image is being displayed to the user, otherwise it looks like a bug.  The point of the blacklist filtering is to avoid showing people stuff that they don't want to see, not prevent them from saving it once they've bypassed the blacklist to look at it anyway.

I guess the case I was thinking of is where a user follows a direct link, maybe from another site, to an individual e621 submission.  If the user hasn't disabled the blacklist on e621 yet, they'll get the e621 submission page, with all the tags, description, etc, but there will be a "blacklisted" placeholder image where the real image should be.  The user then has to click the "disable all" link on the submission page to get the real image to show up.

I was wondering if we need to detect that placeholder image (or other indication of blacklisted-ness) on the individual submission pages, and either run or not run the plugin (raccoon-head appears or not) as appropriate.

History:

I *thought* that the blacklist-disable on e621 used to not "stick" from the search-results page to the individual submission pages.  In other words, if you ran a search, got results with some submissions blacklisted, and clicked "disable all" on the search page, the thumbnails for the previously-blacklisted submissions would pop in on the search page.  BUT, if you clicked one of those newly-visible thumbnails, you'd get the submission page with the "blacklisted" placeholder image, and you'd have to click "disable all" again on the submission page to see the image.

I thought I had seen this behavior when I was testing this change a couple of weeks ago, but I tried again, and this is not (now?) the case.  Once you hit "disable all" on the search page, you don't get the placeholder images on the individual submission pages either.

> It might make sense to generate an initial list and then check for the blacklisted class on the [element.classList property](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList) on that initial list of elements and filter them out if it's present.

The difficulty I see with doing that is that I still need some way to know when the blacklisting script is done.  In other words, if I generate a list of submissions, and then look to see if anything on that list has the `.blacklisted-active` in its class, I can get five possible results, only two of which are desirable:

`.blacklisted-active` not found:
1. The blacklist script hasn't ran at all yet.
2. Blacklist script has run for at least *some* submissions, but it hasn't found anything to blacklist yet.  It might not have run on all submissions yet, so I might end up with a race condition.
3. Blacklist script has run for all submissions, but it didn't find anything to blacklist, so it didn't modify any of the classes.  It is OK to proceed with the list I have.

`.blacklisted-active` found:

4. Blacklist script has run for at least *some* submissions, and blacklisted at least one submission.  It might not have run on all submissions yet, so I might end up with a race condition.
5. Blacklist script has run for all submissions.  It is OK to remove the blacklisted submissions from my list.

My initial proposed fix used the presence of the `.blacklisted` tag to rule out case 1, and just hoped that cases 2 and 4 didn't happen.  Since the presence of the `.blacklisted` tag seems to not be reliable, my proposed fix doesn't always work.

I *might* be able to use the presence of the "blacklist-box"/"sidebar-blacklist" in the side bar to decide if the blacklist script is done - not sure yet.  This may still lead me back to the case where if it isn't there, I can't tell the difference between "hasn't run yet" and "it ran and nothing is blacklisted".

There is another wrinkle - a possible change at e621 - which I will mention in my next reply.

**Comment by Eupeptic1** [comment]
**tl;dr:** There is a (possible) upcoming change at e621 that may affect the fix for this issue.

I *think*, but can't *prove*, that e621 might be behaving differently than when I wrote my initial fix for this.  Investigating, I eventually found [this change](https://github.com/zwagoth/e621ng/commit/765c6769fb322fe7f2f4314e1b3c596b4e2e533d) to e621's blacklist code, from about six weeks ago.  (Top level for that repo [here](https://github.com/zwagoth/e621ng) .)

I don't *think* that change has been deployed to production e621 yet.  In some cases, I still see the `.blacklisted` ("blacklist has run") and `.blacklisted-active` ("submission is blacklisted") classes.  I *think* that change does away with `.blacklisted-active` completely, and repurposes `.blacklisted` to mean "submission is blacklisted".

I don't know, and probably can't find out, what e621's deployment schedule might be for this change.  But if the fix for the Raccoony issue depends on how the e621 blacklist currently works, it might break when that e621 change is deployed.

**Comment by Simon-Tesla** [comment]
Huh, my recollection was that blacklisting didn't hide the images on individual pages. If e621 is showing a placeholder, I'd say whatever is easiest to implement is fine.

**Comment by Eupeptic1** [comment]
Just as a heads-up, I am probably not going to have much time to work on this until early February or so.

(I notice you are merging some other pull requests, which sometimes means a release is soon, so I wanted to provide an update.)

**Comment by Simon-Tesla** [comment]
No worries. Not sure on a timeline for next release, just kinda trying to make some progress, even if it's small.

**Comment by Eupeptic1** [comment]
**tl;dr:** I took another pass at it and it seems to be working OK for me.  Proposed change is on the [e621-blacklist](https://github.com/Eupeptic1/RaccoonyWebEx/tree/e621-blacklist) branch in my repo; only the e621 plugin itself changed.

#### Changes from previous proposed fix

The change at e621 I mentioned [previously](https://github.com/Simon-Tesla/RaccoonyWebEx/issues/122#issuecomment-1003659303) seems to have been put in production.  Now, only blacklisted submissions have `.blacklisted` in their class.  `.blacklist-active` doesn't appear to be used any longer.

This actually simplifies the plugin, when used in conjunction with the mutation selector.  Now, I just look for everything that doesn't have `.blacklisted` in the class, and trust the mutation selector to call me after e621's client-side blacklist script has run.

#### Testing

It seems to work OK, in limited testing. If there is a mix of non-blacklisted and blacklisted submissions on the page, the blacklisted submissions do not appear in the open-in-tabs list. If all of the submissions on the page are blacklisted, the raccoon head does not appear at all. (Try being logged out and searching a tag on e621's default blacklist to see this.)

If one clicks the "disable all" link under the "Blacklisted" heading in the left side bar, the plugin runs again, and includes the newly-visible posts in the open-in-tabs list, along with the posts that were previously visible.

If nothing on the page is blacklisted, everything on the page makes it to the open-all-in-tabs list.

It works the same when you are not logged in (tested with tags on the default blacklist) as when you are logged in (tested with tags on my personal blacklist).

It works OK on [this search](https://github.com/Simon-Tesla/RaccoonyWebEx/issues/122#issuecomment-1003546679), which previously didn't work correctly.

If you follow a direct link to a single blacklisted submission, Raccoony still offers the raccoon-head menu, and Raccoony's Download button works, regardless of whether e621 is displaying the "Blacklisted" placeholder image or not.

#### Notes

If you are looking at a search results page on e621, and then type more than three letters into the search box at e621, the plugin gets called again for every letter you type, as e621 modifies the page to show possible tags under the search box.  I'm pretty sure this is the mutation selector at work.

If you decide not to make a new search, and click out of the search box, the plugin still finds the same set of links it previously found, so the "open in tabs" button still does what you would expect.

The repeated plugin calls while you are typing in the search box don't seem to be a performance problem on my machine (2.9 GHz, dual-core, 4 GB), but I figured I would note it.

#### Availability

The revised proposed change is on the [e621-blacklist](https://github.com/Eupeptic1/RaccoonyWebEx/tree/e621-blacklist) branch in my repo.  Only the e621 plugin itself changed.

**Comment by Simon-Tesla** [comment]
Now that I'm moved and life is starting to (hopefully) settle down, looking at doing some work on Raccoony again. Did you want to open a pull request on this?

**Comment by Eupeptic1** [comment]
I've been dogfooding this since February, but I haven't specifically been checking that the proposed change DTRT for blacklisted submissions.  I know that the basic functionality of open-all-in-tabs and downloading individual submissions still works correctly on e621.

I am going to do one more round of testing, including manual and semi-automated, in at least Chrome and Firefox, and then open a pull request.  I expect this to happen by Friday, 3 June 2022.

**Comment by Eupeptic1** [comment]
I did the manual testing and semi-automated testing in Chrome and Firefox.  The tests were successful.

I will submit a pull request shortly.

## Test report 

### Versions tested

Raccoony: 1.2.5 + this fix only

Browsers:
Firefox 91.9.0esr 64-bit, desktop
Chrome 102.0.5005.61 64-bit, desktop
both browsers on Debian Linux 10.12, 64-bit, desktop

### Test of this fix, while logged out of e621

I searched a tag that is on e621's default blacklist.  e621 gave a search results page with no thumbnails.  Raccoony detected no submissions, as seen in the console.  The raccoon-head menu did not appear.  This was the expected result.

I searched a tag that is not on e621's default blacklist, but many of the submissions with that tag also have a tag that is on e621's default blacklist.  e621 gave a search results page with 10 thumbnails.  Raccoony detected those 10 submissions only, as seen in the console.  The raccoon-head menu appeared.  Clicking Raccoony's "open all in tabs" button opened those 10 submissions only.  This was the expected result.

I searched a tag that is not on e621's default blacklist, and that should only return 2 submissions.  e621 gave a search results page with 2 thumbnails.  Raccoony detected those 2 submissions only, as seen in the console.  The raccoon-head menu appeared. Clicking Raccoony's "open all in tabs" button opened those 2 submissions only.  This was the expected result.

### Test of this fix, while logged in to e621

I logged in to e621 on my personal account.

I searched a tag that is on my personal blacklist.  e621 gave a search results page with no thumbnails.  Raccoony detected no submissions, as seen in the console.  The raccoon-head menu did not appear.  This was the expected result.

I searched a tag that is not on my personal blacklist, but some of the submissions with that tag also have a tag that is on my blacklist.  e621 gave a search results page with 70 thumbnails.  Raccoony detected those 70 submissions only, as seen in the console.  The raccoon-head menu appeared.  This was the expected result.

I searched a tag that is not on my personal blacklist, and none of the submissions with that tag also have a tag on my blacklist. e621 gave a search results page with 2 thumbnails.  Raccoony detected those 2 submissions only, as seen in the console.  The raccoon-head menu appeared. Clicking Raccoony's "open all in tabs" button opened those 2 submissions only. This was the expected result.

### Regression testing

I used the bookmarks for e621 only from the 2021-11-21 version of the test bookmark file (provided via private email).

I downloaded all 14 e621 bookmarks under all eight of the following conditions:

FIrefox, 1.2.5, logged out of e621
Firefox, 1.2.5, logged in to e621
Firefox, this fix, logged out of e621
Firefox, this fix, logged in to e621

Chrome, 1.2.5, logged out of e621
Chrome, 1.2.5, logged in to e621
Chrome, this fix, logged out of e621
Chrome, this fix, logged in to e621

All eight sets had 28 files - 14 submissions, and 14 metadata files.  All eight sets compared binary equal with each other.

This shows no regression.
